google-cloud-core/src/main/java/com/google/cloud/ServiceOptions.java

```diff
@@ -213,13 +214,18 @@ public B setHost(String host) {
     }
 
     /**
-     * Sets the service authentication credentials.
+     * Sets the service authentication credentials. If this method or {@link #setNoCredentials() are
+     * not used on the builder, {@link GoogleCredentials#getApplicationDefault()} will be used to
+     * attempt getting credentials from the environment.
      *
+     * @param credentials authentication credentials, should not be {@code null}
      * @return the builder
+     * @throws NullPointerException if {@code credentials} is {@code null}. To disable
+     *     authentication use {@link Builder#setNoCredentials()}
```

Comment:
The `setNoCredentials` pattern feels odd to me. It's a little like the pattern we were previously using in GAX, where different entry points overwrote the same data. I'd prefer one entry point I think (`setCredentials`). Could we create these types?

* `NullCredentials` - no credentials are used; for the emulator case.
* `DefaultCredentials` - I know this feels a little redundant with `GoogleCredentials.getApplicationDefault()`, but maybe there will be a case of a service where ADC is not the right default credentials to use for it. Additionally, having the user call `GoogleCredentials.getApplicationDefault()` themselves is slightly different, because the point of call for credential acquisition is different.